### PR TITLE
Fix sources/artifacts bug introduced with #184

### PR DESCRIPTION
**Pull request by siredmar**
This fixes issue #200. The file handling introduced with #184 relies on absolute paths. 

It checks if the path is relative. If so, prepent the `cwd` to the path.
e.g. 
```
job=build:
    use: builder
    sources: 
      - source/*.cpp
...
```
In the above example the paths in sources are relative. They are located within the project where the dobi.yaml is located. So internally it creates `<cwd>/source/*.cpp`.

### CODE REVIEW

File: utils/fs/directory.go
Comment (dnephin) [inline utils/fs/directory.go:56]: This will always call `os.Getwd()`, right? I think we can call `os.Getwd()` directly here instead of exporting `ValueFromFilesystem`.

File: utils/fs/directory.go
Comment (dnephin) [inline utils/fs/directory.go:62]: Couple thoughts here:

1. I guess this will break on windows, this should probably use https://golang.org/pkg/path/filepath/#IsAbs
2. Since `rootPath` never changes I think we can do this work before the `walker` function instead of inside `walker`.

I believe if `rootPath` is absolute that `filePath` will always be absolute, but a test for each of those cases should confirm.

**Review by dnephin — COMMENTED** [review]
Thank you for tracking down this bug and submitted a fix!

File: utils/fs/directory.go
Comment (siredmar) [reply utils/fs/directory.go:56]: You are absolutely right. Using `os.Getwd()` directly.

File: utils/fs/directory.go
Comment (siredmar) [reply utils/fs/directory.go:62]: 1. yes, i didn't think about that. Using `filepath.IsAbs()` and `os.PathSeperator` 
2. you are right, will move that one outside. 

Btw: the reason why the tests always succeeded with those changes was that `tmpdir := os.NewDir` and `tmpdir.Path()` always returns absolute paths. I will add a directory test for relative paths.

**Comment by siredmar** [comment]
@dnephin would you mind making a release 0.15.0 once this is merge? Thx!

**Comment by siredmar** [comment]
@dnephin this is just a ping

**Comment by siredmar** [comment]
@dnephin this is another ping. Can you merge this please, if is good enough?

**Comment by dnephin** [comment]
Thank you for working on this fix! Sorry it has taken me some time to have a look. I started looking and I found a few more problems. I made some changes here: https://github.com/dnephin/dobi/commit/2aa660834b35f8300dba1f30e020bf96ef96b39b

I noticed that all of the callers to this function should be passing `Root` by getting it from the `ExecuteContext`, but there are a few places that do not.

I think instead of making everything an absolute path , then later making it a relative path, it should keep relative paths as they are, and only use `filepath.Rel` when it receives an absolute path.

I'll see if I can get that working today, but it might still need some work.

**Comment by dnephin** [comment]
I rebased your changes on the latest master and added another commit. Does this work to solve the bug you hit? The tests are passing, but I made some changes to them, and I'm not sure if they still accurately reproduce the problem.

The change I made was to require that `LastModifiedSearch.Root` always be set to an absolute path, and I documented that on the field. All callers should now use an absolute path, so `filepath.Rel` should work.

**Comment by dnephin** [comment]
hmm, it looks like this change to make the context relative to the working directory is not correct. The tests are failing.

**Comment by siredmar** [comment]
@dnephin without investigated the issue the failing circle ci tests might fail because the paths might put `<absolute>/<relative>` together. This is the image that fails:
Example
```
image=builder:
    image: dobi-dev
    context: dockerfiles/
    dockerfile: Dockerfile.build
```
And the error stating `<absolute>/<relative>`  for path:
```
[ERROR] failed to execute task "builder:build": internal error: stat /root/project/dockerfiles/dockerfiles: no such file or directory
```

**Comment by siredmar** [comment]
@dnephin i was right. For the stale check in the image build the root path was false set. The task configuration brings the correct context, therefore no need to append the context to the workingDir.

**Comment by siredmar** [comment]
@dnephin i can confirm with my manual tests that the current state of this branch solves my issues. You might want to merge this now.

**Comment by siredmar** [comment]
@dnephin Can this get merged now?

**Comment by siredmar** [comment]
@dnephin Is this project dead now?

**Comment by siredmar** [comment]
@dnephin I don't want to be a pain, but in the meantime I need a fixed release with an even older feature. So would you please merge this and create a new release `0.15.0`?

File: tasks/image/build.go
Comment (dnephin) [inline tasks/image/build.go:80]: This doesn't seem like it can be right. The last modified time should be based on the context directory, not the working directory.

**Review by dnephin — APPROVED** [review]
Thank you for your help and your patience in fixing this bug!

I think this is working correctly now. I believe you were correct that joining the working directory and context was not exactly correct, it needs to only join them when the context is relative.

**Comment by dnephin** [comment]
https://github.com/dnephin/dobi/releases/tag/v0.15.0

**Pull request merged**
